### DISCUSSION TITLE
Please teach me how to modify micropython code

### DISCUSSION THREAD

**Discussion by carl336** [micropython/micropython#11776] [RP2040 / Pico]
Hello, I would like to modify the MicroPython code, but I have no idea how to modify it so that it runs like this.
1.	First of all, after pressing KEY 1 and immediately releasing it, the following data is executed:
Vdd(1)
time.sleep(0.1)
cs(0)
spi_write_data_1()
spi_write_data_4()
spi_write_data_5()
spi_write_data_3()
spi_write_data_2()
cs(1)
time.sleep(0.008)
2.	Then, if you press KEY2 at first and release it immediately, the following data will be executed:
cs(0)
spi_write_data_6()
cs(1)
3.	Then, if you press KEY2 a second time and release it immediately, this data is executed: cs(0) spi_write_data_6() cs(1)
cs(0)
spi_write_data_7()
cs(1)
4.	No. 2 and 3 can be repeated all the way through
5.	Finally, if KEY1 is pressed and released immediately, this data is executed:
cs(0)
spi_write_data_1()
spi_write_data_2()
spi_write_data_8()
spi_write_data_4()
spi_write_data_5()
cs(1)
Vdd.value(0)
time.sleep(0.1)　
6.	after No. 5 is executed, it returns to No. 1 and can be repeated indefinitely. 
7.	The contents of 1st to 5th are the first loop, where no code is executed before and after the last press of KEY1. In other words, data is output only if KEY1 is pressed first and KEY2 is pressed immediately after.
I will upload the waveform diagram analyzed by Logic Analyzer. Please refer to it.
As shown in the picture,
1.	Channels 0 represent Vdd.
2.	Channel 7 represents KEY1 and channel 8 represents KEY2.
3.	The two parts circled in white in the figure are loops (repetitions) after the code is executed.
4.	The area circled in green is the signal that is generated outside of this cycle (repetition) when button 2 is pressed and immediately released.
5.	I want to eliminate the signal generated from the area circled in green (the same as the area circled in blue)

Here are the codes that I wrote below:

from machine import Pin, SPI
import time

KEY1 = Pin(0, Pin.IN, Pin.PULL_UP)
KEY2 = Pin(15, Pin.IN, Pin.PULL_UP)
Vdd = Pin(2, mode=Pin.OUT, value=0)
spi = SPI(0, baudrate=7500000, polarity=0, phase=0, bits=8, firstbit=SPI.MSB, sck=Pin(6), mosi=Pin(7))
cs = Pin(1, mode=Pin.OUT, value=0)
cs(1)
Vdd(0)
state = False

def spi_write_data_1():
    spi.write(bytes([0x01]))

def spi_write_data_2():
    spi.write(bytes([0x02]))

def spi_write_data_3():
    spi.write(bytes([0x03]))

def spi_write_data_4():
    spi.write(bytes([0x04]))

def spi_write_data_5():
    spi.write(bytes([0x05]))

def spi_write_data_6():
    spi.write(bytes([0x06]))

def spi_write_data_7():
    spi.write(bytes([0x07]))

def spi_write_data_8():
    spi.write(bytes([0x08]))

def reverseGPIO():
    if Vdd.value() == 1:
        cs(0)
        spi_write_data_1()
        spi_write_data_2()
        spi_write_data_8()
        spi_write_data_4()
        spi_write_data_5()
        cs(1)
        time.sleep(0.008)
       
    else:
        Vdd(1)
        time.sleep(0.1)
      
        cs(0)
        spi_write_data_1()
        spi_write_data_4()
        spi_write_data_5()
        spi_write_data_3()
        spi_write_data_2()
        cs(1)
        
        time.sleep(0.008)

try:
    while True:
        if KEY1.value() == 0:
            time.sleep_ms(10)
            if KEY1.value() == 0:
                reverseGPIO()

                while KEY1.value() == 0:
                    time.sleep_ms(10)

        if KEY2.value() == 0:
            state = not state

            if state:
                cs(0)
                spi_write_data_6()
                cs(1)
                
              
            else:
                cs(0)
                spi_write_data_7()
                cs(1)

            while KEY2.value() == 0:
                time.sleep_ms(10)

except:
    pass

<img width="960" alt="图片_20230613172148" src="https://github.com/micropython/micropython/assets/133198158/c8df95a4-efa6-434f-a840-75790e8a07ba">

**Comment by ned-pcs** [comment]
You should first draw a state diagram that documents your desired response to the different keystrokes. This will guide your code. You will need to implement a simple state machine that provides the behavior in your state diagram.

The SPI output of steps 1, 2, 3, and 5 could be mapped directly into states, and the keypresses mapped into transitions between those states.

**Comment by jimmo** [comment]
Please also see https://github.com/orgs/micropython/discussions/9111 about how to format code in posts.

State machines are definitely the way to go here.
